diff --git libs/icecream-cpp/icecream.hpp libs/icecream-cpp/icecream.hpp
index c01261ec2..d8ae65a83 100644
--- libs/icecream-cpp/icecream.hpp
+++ libs/icecream-cpp/icecream.hpp
@@ -97,21 +97,21 @@
     #endif
 #endif
 
-#if defined(__has_include) && __has_include(<ranges>)
+#if defined(__has_include) && __has_include(<ranges>) && (__cplusplus >= 201703L)
     #include <ranges>
     #if defined(__cpp_lib_ranges)
         #define ICECREAM_LIB_RANGES
     #endif
 #endif
 
-#if !defined(__APPLE__) && defined(__has_include) && __has_include(<source_location>)
+#if !defined(__APPLE__) && defined(__has_include) && __has_include(<source_location>) && (__cplusplus >= 201703L)
     #include <source_location>
     #if defined(__cpp_lib_source_location)
         #define ICECREAM_SOURCE_LOCATION
     #endif
 #endif
 
-#if defined(__has_include) && __has_include(<format>)
+#if defined(__has_include) && __has_include(<format>) && (__cplusplus >= 201703L)
     #include <format>
     #if defined(__cpp_lib_format) || (defined(_LIBCPP_VERSION) && _LIBCPP_VERSION >= 170000 && __cplusplus >= 202002L)
         // libc++ just defines the '__cpp_lib_format' macro start from version 19. However
@@ -129,7 +129,7 @@
 // The "defined()" test for FMT_VERSION macro is an attempt to check if any {fmt} header
 // was "#included" before the including of this Icecream-cpp header. If so, the {fmt}
 // support will be enabled.
-#if defined(ICECREAM_FMT) || defined(FMT_VERSION)
+#if (defined(ICECREAM_FMT) || defined(FMT_VERSION)) && (__cplusplus >= 201703L)
     // All the {fmt} headers from supoorted versions (5.0 and newer) will directly or
     // indirectly include a source file where FMT_VERSION is defined.
     #include <fmt/format.h>
@@ -2193,8 +2193,8 @@ namespace icecream{ namespace detail
         auto operator()(StringView str) -> void
         {
             static_assert(
-                sizeof(std::streamsize) <= sizeof(size_t),
-                "std::streamsize must not be wider than size_t"
+                sizeof(std::streamsize) <= sizeof(long long),
+                "std::streamsize must not be wider than long long"
                 "Please report a bug on https://github.com/renatoGarcia/icecream-cpp/issues"
             );
 
@@ -2206,7 +2206,7 @@ namespace icecream{ namespace detail
                 // This will handle the unlikely situation where a huge str is big
                 // enough to its size overflow when converted to a `std::streamsize`
                 auto chunk_size =
-                    (remaining > static_cast<size_t>(std::numeric_limits<std::streamsize>::max()))
+                    (remaining > static_cast<long long>(std::numeric_limits<std::streamsize>::max()))
                         ? std::numeric_limits<std::streamsize>::max()
                         : static_cast<std::streamsize>(remaining);
 
@@ -5404,7 +5404,7 @@ namespace detail {
             , arg_names_(arg_names)
         {}
 
-        // Used by compilers that expand an empyt __VA_ARGS__ in
+        // Used by compilers that expand an empty __VA_ARGS__ in
         // Dispatcher{bla, #__VA_ARGS__} to Dispatcher{bla, }
         Dispatcher(
             bool is_ic_apply,
